addons/portal/controllers/portal.py

```diff
@@ -230,6 +232,325 @@ def account(self, redirect=None, **post):
+        if 'state_id' not in address_values and partner_sudo.state_id:
+            address_values['state_id'] = partner_sudo.state_id.id
+
+        # Validate the address values and highlights the problems in the form, if any.
+        invalid_fields, missing_fields, error_messages = self._validate_address_values(
+            address_values,
+            partner_sudo,
+            address_type=address_type,
+            **extra_form_data,
+        )
+
+        if error_messages:
+            return partner_sudo, {
+                'invalid_fields': list(invalid_fields | missing_fields),
+                'messages': error_messages,
+            }
+        if not partner_sudo:  # Creation of a new address.
+            address_values['parent_id'] = address_values.get('parent_id', request.env.user.partner_id.id)
+            create_context = tools.clean_context(request.env.context)
+            create_context.update({
+                'tracking_disable': True,
+                'no_vat_validation': True,
+            })
+            partner_sudo = request.env['res.partner'].sudo().with_context(
+                create_context
+            ).create(address_values)
+        else:
+            partner_sudo.write(address_values)
+        callback = form_data.get('callback') or '/my/addresses'
+        return partner_sudo, {'successUrl': callback}
+
+    # moved from website_sale
+    def _parse_form_data(self, form_data):
+        """ Parse the form data and return them converted into address values and extra form data.
+        :param dict form_data: The form data to convert to address values.
+        :return: A tuple of converted address values and extra form data.
+        :rtype: tuple[dict, dict]
+        """
+        address_values = {}
+        extra_form_data = {}
+
+        ResPartner = request.env['res.partner']
+        partner_fields = ResPartner._fields
+        for key, value in form_data.items():
+            if isinstance(value, str):
+                value = value.strip()
+            if key in partner_fields:
+                field = partner_fields[key]
+                if field.type == 'many2one' and isinstance(value, str) and value.isdigit():
+                    address_values[key] = field.convert_to_cache(int(value), ResPartner)
+                else:
+                    # Always keep field values, even if falsy, as it might be for resetting a field.
+                    address_values[key] = field.convert_to_cache(value, ResPartner)
+            elif value:  # The value cannot be saved on the `res.partner` model.
+                extra_form_data[key] = value
+        if (
+            hasattr(ResPartner, 'check_vat')  # The `base_vat` module is installed.
+            and address_values.get('vat')
+            and address_values.get('country_id')
+        ):
+            address_values['vat'] = ResPartner.fix_eu_vat_number(
+                address_values['country_id'],
+                address_values['vat'],
+            )
```

Comment:
I was wondering if this code snippet could be an override or something similar but I think we're limited in this controller scope :thinking: